### PR TITLE
Actual log rate computing and bad rpm values filtration for spectrum chart.  

### PR DESCRIPTION
**Pull request by demvlad**
My tinywhoops Happymodel F411 flight controller has 1/3 data lost by writing flight data to Highspeed Openlagger serial device with 2k log rate.
The BlackBoxExplorer shows wrong frequency at the spectral
diagramm. The reasone this is the big difference between
actual log rates and Betaflight value what BBE calculates by configuration data.
This PR resolved this issue.
It calculates actual log rate once, and compares it with config values.
If the difference is above 5%, then the FFT algoritm uses the actual rate value. The warning message is showed at the spectr chart in this case.
I've tryed to use actual rate allways without condition and  checked this at my olde good logs from 5" quad. It had about 2002Hz actual rate instead of 2000. Therefore i use actual rates in case of relative big difference only - 5%.
This PR will be helpfull to users, who have happymodel tinywhoops with High speed OpenLagger serial blackbox.
Thank's @mituritsyn for log file samples count computing algoritm.

### CODE REVIEW

File: js/graph_spectrum.js [hunk 1]
```diff
@@ -125,13 +125,13 @@ var
...
             }
-
+            return sampleInfo;
         };
 
         /* This function is called from the canvas drawing routines within grapher.js
            It is only used to record the current curve positions, collect the data and draw the 
            analyser on screen*/
-        this.plotSpectrum =	function (fieldIndex, curve, fieldName) {
+        this.plotSpectrum =    function (fieldIndex, curve, fieldName) {
```
Comment (nerdCopter) [inline js/graph_spectrum.js:134]:
```suggestion
        this.plotSpectrum = function (fieldIndex, curve, fieldName) {
```

File: js/graph_spectrum.js [hunk 2]
```diff
@@ -142,8 +142,10 @@ var
             // Detect change of selected field.... reload and redraw required.
             if ((fftData == null) || (fieldIndex != fftData.fieldIndex) || dataReload) {
                 dataReload = false;
-                dataLoad();			
+                const logRateInfo = dataLoad();
```
Comment (nerdCopter) [inline js/graph_spectrum.js:145]:
```suggestion
                const logRateInfo = dataLoad();
```

File: js/graph_spectrum_calc.js [hunk 3]
```diff
@@ -51,6 +54,33 @@ GraphSpectrumCalc.setOutTime = function(time) {
 
 GraphSpectrumCalc.setDataBuffer = function(dataBuffer) {
     this._dataBuffer = dataBuffer;
+
+    if (this._actualeRate == undefined) {
```
Comment (nerdCopter) [inline js/graph_spectrum_calc.js:58]:
```suggestion
    if (this._actualeRate == undefined) {
```

File: js/graph_spectrum_calc.js [hunk 4]
```diff
@@ -51,6 +54,33 @@ GraphSpectrumCalc.setOutTime = function(time) {
 
 GraphSpectrumCalc.setDataBuffer = function(dataBuffer) {
     this._dataBuffer = dataBuffer;
+
+    if (this._actualeRate == undefined) {        
+        const minTime =  this._analyserTimeRange.in;
+        const maxTime =  this._analyserTimeRange.out;
+        this._analyserTimeRange.in = this._flightLog.getMinTime();
```
Comment (nerdCopter) [inline js/graph_spectrum_calc.js:61]:
```suggestion
        this._analyserTimeRange.in = this._flightLog.getMinTime();
```

File: js/graph_spectrum_calc.js [hunk 5]
```diff
@@ -51,6 +54,33 @@ GraphSpectrumCalc.setOutTime = function(time) {
...
+        if(timeRange > MAX_ANALYSER_LENGTH) {
+            timeRange = MAX_ANALYSER_LENGTH;
+            this._analyserTimeRange.out = this._analyserTimeRange.in + MAX_ANALYSER_LENGTH;
+        }
+        const data = this._getFlightSamplesFreq();
+        this._actualeRate = 1e6 * data.count / timeRange;
+        if (Math.abs(this._BetaflightRate - this._actualeRate) / this._actualeRate > WARNING_RATE_DIFFERENCE)
+            this._blackBoxRate = Math.round(this._actualeRate);
+
+        this._analyserTimeRange.in = minTime;
```
Comment (nerdCopter) [inline js/graph_spectrum_calc.js:73]:
```suggestion
        this._analyserTimeRange.in = minTime;
```

**Comment by nerdCopter** [comment]
recommend setting your editor to auto-trim trailing spaces. and use 4-sapces for tabs automatically.

**Comment by demvlad** [comment]
> recommend setting your editor to auto-trim trailing spaces. and use 4-sapces for tabs automatically.

Ok. I have 4-space auto.
But autotrim.. Thank's it is interest. Wiil find..

**Comment by demvlad** [comment]
But i think that the warning about this situation must be at the main program window, it is not at internal spectrum chart only.

File: js/graph_spectrum_plot.js [hunk 6]
```diff
@@ -855,3 +858,30 @@ GraphSpectrumPlot._invalidateCache = function() {
...
+              betaflightRate = this._logRateWarning.betaflightRate.toFixed(0);
+        const WarningText = "THE ACTUAL AND CONFIG LOG DATA RATE DIFFERENCE: " + actualRate + " : " + betaflightRate;
+        const X = canvasCtx.canvas.width/2,
+              Y = canvasCtx.canvas.height/12;
+        canvasCtx.strokeText(WarningText, X, Y);
+        canvasCtx.fillText(WarningText, X, Y);
+
+        canvasCtx.restore();
+    }
+};
```
Comment (haslinghuis) [inline js/graph_spectrum_plot.js:887]:
```suggestion
};

```

**Comment by demvlad** [comment]
The last changes improves issue, what i found on my Happymodel FC.
There are very big wrong RPM values  ever in the logs.
This improvement resolved this.

File: js/graph_spectrum_plot.js [hunk 7]
```diff
@@ -855,3 +858,30 @@ GraphSpectrumPlot._invalidateCache = function() {
 GraphSpectrumPlot._invalidateDataCache = function() {
     this._cachedDataCanvas = null;
 };
+
+GraphSpectrumPlot.setLogRateWarningInfo  = function(logRateInfo) {
+    this._logRateWarning = logRateInfo;
+};
+
```
Comment (haslinghuis) [inline js/graph_spectrum_plot.js:865]: :space_invader:

File: js/graph_spectrum_calc.js [hunk 8]
```diff
@@ -25,14 +27,38 @@ var GraphSpectrumCalc = GraphSpectrumCalc || {
...
+            this._blackBoxRate = Math.round(this._actualeRate);
+
+    if (this._BetaflightRate != this._blackBoxRate)
+        return {
+            actualRate: this._actualeRate,
+            betaflightRate: this._BetaflightRate
+        };
+    else
+        return undefined;
+
```
Comment (haslinghuis) [inline js/graph_spectrum_calc.js:61]:
```suggestion
    if (this._BetaflightRate !== this._blackBoxRate) {
        return {
            actualRate: this._actualeRate,
            betaflightRate: this._BetaflightRate
        };
    }
    return undefined;
```

File: js/graph_spectrum.js [hunk 9]
```diff
@@ -75,7 +75,7 @@ var
             }
         };
 
-       	this.resize = function() {
+           this.resize = function() {
```
Comment (haslinghuis) [inline js/graph_spectrum.js:78]:
```suggestion
       this.resize = function() {
```

**Review by haslinghuis — CHANGES_REQUESTED** [review]
Did not check them all - as to minimize refactoring efforts.

File: js/graph_spectrum_plot.js (hunk 7 above)
Comment (haslinghuis) [inline js/graph_spectrum_plot.js:865]:
```suggestion
```

File: js/graph_spectrum_calc.js [hunk 10]
```diff
@@ -25,14 +27,37 @@ var GraphSpectrumCalc = GraphSpectrumCalc || {
...
+    const length = allChunks.reduce((acc, chunk) => acc + chunk.frames.length, 0);
+    this._actualeRate = 1e6 * length / timeRange;
+
+     if (Math.abs(this._BetaflightRate - this._actualeRate) / this._actualeRate > WARNING_RATE_DIFFERENCE)
+            this._blackBoxRate = Math.round(this._actualeRate);
+
+    if (this._BetaflightRate !== this._blackBoxRate) {
+        return {
+            actualRate: this._actualeRate,
+            betaflightRate: this._BetaflightRate
```
Comment (haslinghuis) [inline js/graph_spectrum_calc.js:57]:
```suggestion
            betaflightRate: this._BetaflightRate,
```

**Review by haslinghuis — APPROVED** [review]

**Review by nerdCopter — APPROVED** [review]
- approving based on conversations and still-working status with a few logs. did not deep-test.

**Pull request merged**